src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Completion/CompletionTriggerAndCommitCharacters.cs

```diff
@@ -0,0 +1,117 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.Razor.Workspaces;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.Razor.Completion;
+
+internal class CompletionTriggerAndCommitCharacters
+{
+    /// <summary>
+    ///  Trigger character that can trigger both Razor and Delegation completion
+    /// </summary>
+    private const char TransitionCharacter = '@';
+
+    private static readonly char[] s_vsHtmlTriggerCharacters = [':', '#', '.', '!', '*', ',', '(', '[', '-', '<', '&', '\\', '/', '\'', '"', '=', ':', ' ', '`'];
+    private static readonly char[] s_vsCodeHtmlTriggerCharacters = ['#', '.', '!', ',', '-', '<'];
+    private static readonly char[] s_razorTriggerCharacters = ['<', ':', ' '];
+    private static readonly char[] s_csharpTriggerCharacters = [' ', '(', '=', '#', '.', '<', '[', '{', '"', '/', ':', '~'];
+    private static readonly char[] s_commitCharacters = [' ', '>', ';', '='];
+
+    private readonly HashSet<char> _csharpTriggerCharacters;
+    private readonly HashSet<char> _delegationTriggerCharacters;
+    private readonly HashSet<char> _htmlTriggerCharacters;
+    private readonly HashSet<char> _razorTriggerCharacters;
+    private readonly ImmutableArray<string> _allTriggerCharacters;
+    private readonly ImmutableArray<string> _allCommitCharacters;
```

Comment:
❓ is `ImmutableHashSet` faster on lookup than `ImmutableArray`?

Reply:
Oh, and to be clear, `ImmutableHashSet` would absolutely be faster than `ImmutableArray` for lookup for sufficiently large sets. :smile: For small arrays (say a few elements), `ImmutableArray` would be faster. And, `ImmutableHashSet` takes up significantly more memory. 😞